update/sync/syncMiniBlocks.go

```diff
@@ -93,54 +93,72 @@ func (p *pendingMiniBlocks) SyncPendingMiniBlocksFromMeta(
 		listPendingMiniBlocks = append(listPendingMiniBlocks, computedPending...)
 	}
 
-	return p.syncMiniBlocks(listPendingMiniBlocks, waitTime)
+	return p.syncMiniBlocks(listPendingMiniBlocks, ctx)
 }
 
 // SyncPendingMiniBlocks will sync the miniblocks for the given epoch start meta block
-func (p *pendingMiniBlocks) SyncPendingMiniBlocks(
-	miniBlockHeaders []block.ShardMiniBlockHeader,
-	waitTime time.Duration,
-) error {
-	return p.syncMiniBlocks(miniBlockHeaders, waitTime)
+func (p *pendingMiniBlocks) SyncPendingMiniBlocks(miniBlockHeaders []block.ShardMiniBlockHeader, ctx context.Context) error {
+	return p.syncMiniBlocks(miniBlockHeaders, ctx)
 }
 
-func (p *pendingMiniBlocks) syncMiniBlocks(listPendingMiniBlocks []block.ShardMiniBlockHeader, waitTime time.Duration) error {
-	_ = process.EmptyChannel(p.chReceivedAll)
+func (p *pendingMiniBlocks) syncMiniBlocks(listPendingMiniBlocks []block.ShardMiniBlockHeader, ctx context.Context) error {
+	_ = core.EmptyChannel(p.chReceivedAll)
 
-	requestedMBs := 0
-	p.mutPendingMb.Lock()
-	p.stopSyncing = false
+	mapHashesToRequest := make(map[string]uint32)
 	for _, mbHeader := range listPendingMiniBlocks {
-		p.mapHashes[string(mbHeader.Hash)] = struct{}{}
-		miniBlock, ok := p.getMiniBlockFromPoolOrStorage(mbHeader.Hash)
-		if ok {
-			p.mapMiniBlocks[string(mbHeader.Hash)] = miniBlock
-			continue
-		}
-
-		requestedMBs++
-		p.requestHandler.RequestMiniBlock(mbHeader.SenderShardID, mbHeader.Hash)
+		mapHashesToRequest[string(mbHeader.Hash)] = mbHeader.SenderShardID
 	}
+
+	p.mutPendingMb.Lock()
+	p.stopSyncing = false
 	p.mutPendingMb.Unlock()
 
-	var err error
-	defer func() {
+	for {
```

Comment:
duplicate code with syncHeadersByHash.go. Add a TODO to extract the common functionality

Reply:
namings are different. mutexes are different. not really duplicated code - yes, same principle. the code is kept as clean as possible. the thing is underlying structures have different elements - one headers other miniblocks. go does not support generics.

Reply:
Will think of a solution to this.